File: frontend-X2/tool/mock_server.dart
import 'dart:convert';
import 'dart:io';

void main() async {
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 8080);
  print('Mock server listening on http://localhost:8080');

  final usersByEmail = <String, Map<String, dynamic>>{};
  final tokens = <String, Map<String, dynamic>>{};

  await for (HttpRequest request in server) {
    // CORS
    request.response.headers.add('Access-Control-Allow-Origin', '*');
    request.response.headers.add('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    request.response.headers.add('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization');

    if (request.method == 'OPTIONS') {
      request.response.statusCode = HttpStatus.noContent;
      await request.response.close();
      continue;
    }

    final path = request.uri.path;
    try {
      if (path == '/api/v1/auth/register' && request.method == 'POST') {
        final body = await utf8.decoder.bind(request).join();
        final data = jsonDecode(body) as Map<String, dynamic>;
        final email = data['email'] as String? ?? '';
        final password = data['password'] as String? ?? '';
        final role = data['role'] as String? ?? 'CLIENT';

        if (email.isEmpty || password.isEmpty) {
          request.response.statusCode = HttpStatus.badRequest;
          request.response.write(jsonEncode({'message': 'Email et mot de passe requis'}));
          await request.response.close();
          continue;
        }

        if (usersByEmail.containsKey(email)) {
          request.response.statusCode = HttpStatus.conflict;
          request.response.write(jsonEncode({'message': 'Utilisateur déjà existant'}));
          await request.response.close();
          continue;
        }

        final user = {'id': usersByEmail.length + 1, 'email': email, 'role': role};
        usersByEmail[email] = {'password': password, 'user': user};
        final token = 'token-${email.hashCode}-${DateTime.now().millisecondsSinceEpoch}';
        tokens[token] = user;

        request.response.statusCode = HttpStatus.ok;
        request.response.write(jsonEncode({'token': token, 'user': user}));
        await request.response.close();
        continue;
      }

      if (path == '/api/v1/auth/login' && request.method == 'POST') {
        final body = await utf8.decoder.bind(request).join();
        final data = jsonDecode(body) as Map<String, dynamic>;
        final email = data['email'] as String? ?? '';
        final password = data['password'] as String? ?? '';

        final stored = usersByEmail[email];
        if (stored == null || stored['password'] != password) {
          request.response.statusCode = HttpStatus.unauthorized;
          request.response.write(jsonEncode({'message': 'Email ou mot de passe incorrect'}));
          await request.response.close();
          continue;
        }

        final user = stored['user'] as Map<String, dynamic>;
        final token = 'token-${email.hashCode}-${DateTime.now().millisecondsSinceEpoch}';
        tokens[token] = user;

        request.response.statusCode = HttpStatus.ok;
        request.response.write(jsonEncode({'token': token, 'user': user}));
        await request.response.close();
        continue;
      }

      if (path == '/api/v1/auth/logout' && request.method == 'POST') {
        final auth = request.headers.value('authorization');
        if (auth != null && auth.startsWith('Bearer ')) {
          final token = auth.substring(7);
          tokens.remove(token);
        }
        request.response.statusCode = HttpStatus.ok;
        request.response.write(jsonEncode({'message': 'Logged out'}));
        await request.response.close();
        continue;
      }

      if (path == '/api/v1/auth/me' && request.method == 'GET') {
        final auth = request.headers.value('authorization');
        if (auth == null || !auth.startsWith('Bearer ')) {
          request.response.statusCode = HttpStatus.unauthorized;
          request.response.write(jsonEncode({'message': 'Unauthorized'}));
          await request.response.close();
          continue;
        }
        final token = auth.substring(7);
        final user = tokens[token];
        if (user == null) {
          request.response.statusCode = HttpStatus.unauthorized;
          request.response.write(jsonEncode({'message': 'Unauthorized'}));
          await request.response.close();
          continue;
        }
        request.response.statusCode = HttpStatus.ok;
        request.response.write(jsonEncode(user));
        await request.response.close();
        continue;
      }

      // Not found
      request.response.statusCode = HttpStatus.notFound;
      request.response.write(jsonEncode({'message': 'Not found'}));
      await request.response.close();
    } catch (e, s) {
      stderr.writeln('Error handling request: $e\n$s');
      request.response.statusCode = HttpStatus.internalServerError;
      request.response.write(jsonEncode({'message': 'Internal server error'}));
      await request.response.close();
    }
  }
}

File: tool/mock_server.py
#!/usr/bin/env python3
import json
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse
import threading

HOST = '127.0.0.1'
PORT = 8080

users_by_email = {}
tokens = {}

class Handler(BaseHTTPRequestHandler):
    def _set_headers(self, status=200):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization')
        self.end_headers()

    def do_OPTIONS(self):
        self._set_headers(204)

    def do_POST(self):
        parsed = urlparse(self.path)
        length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(length).decode('utf-8') if length else ''
        try:
            data = json.loads(body) if body else {}
        except Exception:
            data = {}

        if parsed.path == '/api/v1/auth/register':
            email = data.get('email', '')
            password = data.get('password', '')
            role = data.get('role', 'CLIENT')
            if not email or not password:
                self._set_headers(400)
                self.wfile.write(json.dumps({'message':'Email et mot de passe requis'}).encode())
                return
            if email in users_by_email:
                self._set_headers(409)
                self.wfile.write(json.dumps({'message':'Utilisateur déjà existant'}).encode())
                return
            user = {'id': len(users_by_email)+1, 'email': email, 'role': role}
            users_by_email[email] = {'password': password, 'user': user}
            token = f"token-{hash(email)}-{__import__('time').time()}"
            tokens[token] = user
            self._set_headers(200)
            self.wfile.write(json.dumps({'token': token, 'user': user}).encode())
            return

        if parsed.path == '/api/v1/auth/login':
            email = data.get('email', '')
            password = data.get('password', '')
            stored = users_by_email.get(email)
            if not stored or stored.get('password') != password:
                self._set_headers(401)
                self.wfile.write(json.dumps({'message':'Email ou mot de passe incorrect'}).encode())
                return
            user = stored['user']
            token = f"token-{hash(email)}-{__import__('time').time()}"
            tokens[token] = user
            self._set_headers(200)
            self.wfile.write(json.dumps({'token': token, 'user': user}).encode())
            return

        if parsed.path == '/api/v1/auth/logout':
            auth = self.headers.get('Authorization','')
            if auth.startswith('Bearer '):
                token = auth[7:]
                tokens.pop(token, None)
            self._set_headers(200)
            self.wfile.write(json.dumps({'message':'Logged out'}).encode())
            return

        self._set_headers(404)
        self.wfile.write(json.dumps({'message':'Not found'}).encode())

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == '/api/v1/auth/me':
            auth = self.headers.get('Authorization','')
            if not auth.startswith('Bearer '):
                self._set_headers(401)
                self.wfile.write(json.dumps({'message':'Unauthorized'}).encode())
                return
            token = auth[7:]
            user = tokens.get(token)
            if not user:
                self._set_headers(401)
                self.wfile.write(json.dumps({'message':'Unauthorized'}).encode())
                return
            self._set_headers(200)
            self.wfile.write(json.dumps(user).encode())
            return
        self._set_headers(404)
        self.wfile.write(json.dumps({'message':'Not found'}).encode())

def run_server():
    server = HTTPServer((HOST, PORT), Handler)
    print(f"Mock server listening at http://{HOST}:{PORT}")
    server.serve_forever()

if __name__ == '__main__':
    run_server()
